keps/sig-cluster-lifecycle/draft-20180119-cluster-api.md

```diff
@@ -0,0 +1,241 @@
+* Declarative Kubernetes upgrades for the control plane and kubelets
+* Declarative upgrades for node OS images
+* Maintaining consistency across different clusters / clouds
+   * By representing important cluster configuration via declarative objects, operations like “diffing” the configuration of two clusters becomes very straightforward. Also, reconcilers can be written to ensure that important cluster configuration is kept in sync between different clusters by simply copying objects.
+* Cloud adoption / lift and shift / liberation
+
+These use cases are in scope for the project, but post-v1alpha1:
+
+* Server-side node draining
+* Autoscaling
+   * Currently, the OSS cluster autoscaler has the responsibility of determining the right size of the cluster and calling the cloud provider to perform the scaling (supporting every cloud provider directly). Modeling groups of nodes in a declarative way would allow autoscalers to only need to worry about the correct cluster size and error handling when that can’t be achieved (e.g. in the case of stockouts), and then separate cloud controllers can be responsible for creating and deleting nodes to reconcile that state and report any errors encountered.
+* Integration with the Cluster Registry API
+   * Automatically add a new cluster to a registry, support tooling that works across multiple clusters using a registry, delete a cluster from a registry.
+* Supporting other common tooling, like monitoring
+
+These use cases are out of scope entirely:
+
+* Creating arbitrary cloud resources
+
+### Cluster-level API
+
+This level of the Cluster Management API describes the global configuration of a cluster. It should be capable of representing the versioning and configuration of the entire control plane, irrespective of the representation of nodes.
+
+Given the recent efforts of SIG Cluster Lifecycle to make kubeadm the de facto standard toolkit for cloud- and vendor-agnostic cluster initialization, and because kubeadm has [an existing API](https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/apis/kubeadm/v1alpha1/types.go) to define the global configuration for a cluster, it makes sense to coalesce the global portion of the Cluster API with the API used by “kubeadm init” to configure a cluster master.
+
+A current goal is to make these APIs as cloud-agnostic as possible, so that the entire definition of a Cluster could remain reasonably in-sync across different deployments potentially in different cloud providers, which would help enable hybrid usecases where it’s desirable to have key configuration stay in sync across different clusters potentially in different clouds/environments. However, this goal is balanced against making the APIs coherent and useable, which strict separation may harm.
+
+The full types for this API can be seen and were initially discussed in [kube-deploy#306](https://github.com/kubernetes/kube-deploy/pull/306).
+
+### Machine API
```

Comment:
I think it's much clearer at this point that this is an api. The cluster part is a bit fuzzier because we don't have a good story for component configuration or addons yet. Once those start getting added I think that the cluster part will look more like an api, but i could be wrong.